### PR TITLE
Fix Windows Command::env("PATH")

### LINKED ISSUE #87859: `Command.env` not working since nightly-2021-07-05
`Command.env` not work after `nightly-2021-07-05` or commit [`8345538`](https://github.com/rust-lang/rust/commit/8345538fec7aa45fabeb7283707386ac7ed51f5b)(maybe).

Here is my test `main.rs`:
```rust
use std::process::Command;
fn main() {
    Command::new("gcc").env("PATH", "C:\\msys64\\mingw64\\bin").spawn().expect("gcc not found");
}
```
```
> rustup default stable-x86_64-pc-windows-gnu
info: using existing install for 'stable-x86_64-pc-windows-gnu'
info: default toolchain set to 'stable-x86_64-pc-windows-gnu'
  stable-x86_64-pc-windows-gnu unchanged - rustc 1.54.0 (a178d0322 2021-07-26)

> rustc main.rs -o stable.exe

> stable.exe      # gcc found !!
gcc.exe: fatal error: no input files
compilation terminated.
```
```
> rustup default nightly-2021-07-05-x86_64-pc-windows-gnu
info: using existing install for 'nightly-2021-07-05-x86_64-pc-windows-gnu'
info: default toolchain set to 'nightly-2021-07-05-x86_64-pc-windows-gnu'
  nightly-2021-07-05-x86_64-pc-windows-gnu unchanged - rustc 1.55.0-nightly (b3d11f95c 2021-07-04)

> rustc main.rs -o nightly.exe

> nightly.exe      # gcc not found !!
thread 'main' panicked at 'gcc not found: Os { code: 2, kind: NotFound, message: "The system cannot find the file specified." }', main.rs:4:73
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**This also affects `rustc` search for linker after nightly-2021-07-05** (with `link-self-contained=yes` or without mingw installed)

Run the `rustc` in cargo's bin works fine, but run the `rustc` in toolchain's bin show this:
```
error: linker `x86_64-w64-mingw32-gcc` not found
```
https://github.com/rust-lang/rust/blob/996ff2e0a0f911f52bb1de6bdd0cfd5704de1fc9/compiler/rustc_codegen_ssa/src/back/linker.rs#L125

### PR DESCRIPTION
**Pull request by ChrisDenton**
Fixes #87859

### CODE REVIEW

**Comment by Urgau** [comment]
I think this need's to be beta nominated because the [PR](https://github.com/rust-lang/rust/pull/85270) who introduce the bug is the the current beta.

**Review by dtolnay — CHANGES_REQUESTED** [review]
The `impl Borrow<OsStr> for EnvKey` lower in this file needs to be removed. That impl was the root cause. The impl is wrong because the Ord and Eq impls have different behavior between OsStr and EnvKey, which violates:

> In particular `Eq`, `Ord` and `Hash` must be equivalent for borrowed and owned values: `x.borrow() == y.borrow()` should give the same result as `x == y`.

**Comment by ChrisDenton** [comment]
Ok I've removed the borrow and changed `sys_common::process::CommandEnv` to work with `EnvKey` instead of `OsStr`/`OsString`. I wonder if it might ultimately be worth Windows using its own implementation of `CommandEnv`.

**Review by dtolnay — APPROVED** [review]

**Comment by ChrisDenton** [comment]
The failure looks like a spurious timeout or something?

**Comment by ChrisDenton** [comment]
Hm, this seems to have failed downloading. Third time lucky?

**Comment by ehuss** [comment]
@bors retry

Failed to download action 'https://api.github.com/repos/actions/checkout/tarball/5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f'. Error: The operation was canceled.

**Comment by ChrisDenton** [comment]
Yeah, remove needs either an EnvKey or else a Borrow that has the same ordering as EnvKey. I could limit the allocation to right before `self.vars.remove`? That would fix the immediate problem and shouldn't have too much impact.

File: library/std/src/sys/windows/process.rs [hunk 1]
```diff
@@ -100,6 +105,20 @@ impl PartialEq for EnvKey {
...
+}
+impl PartialEq<str> for EnvKey {
+    fn eq(&self, other: &str) -> bool {
+        if self.os_string.len() != other.len() {
+            false
+        } else {
+            self.cmp(&EnvKey::new(other)) == cmp::Ordering::Equal
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ChrisDenton) [reply library/std/src/sys/windows/process.rs:121]: Good points. I could remove those implementations and maybe I should add a `#[cfg(not(windows))]` to `have_changed_path`? So as to guard against the possibility of it accidentally being used in the future, or at least until a longer term fix is made.

**Pull request merged**